### PR TITLE
Move deprecation notices on buildForm methods

### LINKED ISSUE #580: Deprecated ColorType error message
#### Sonata packages

```
$ composer show --latest 'sonata-project/*'
sonata-project/admin-bundle    3.40.3 = 3.40.3 The missing Symfony Admin Generator
sonata-project/block-bundle    3.12.1 = 3.12.1 Symfony SonataBlockBundle
sonata-project/cache           1.0.7  = 1.0.7  Cache library
sonata-project/core-bundle     3.11.2 = 3.11.2 Symfony SonataCoreBundle
sonata-project/datagrid-bundle 2.4.0  = 2.4.0  Symfony SonataDatagridBundle
sonata-project/exporter        1.9.1  = 1.9.1  Lightweight Exporter library
```

#### Symfony packages

```
$ composer show --latest 'symfony/*'
symfony/monolog-bundle     v3.3.0  ! v3.3.1  Symfony MonologBundle
symfony/phpunit-bridge     v3.4.18 ~ v4.1.7  Symfony PHPUnit Bridge
symfony/polyfill-apcu      v1.10.0 = v1.10.0 Symfony polyfill backporting apcu_* functions to lower PHP versions
symfony/polyfill-ctype     v1.10.0 = v1.10.0 Symfony polyfill for ctype functions
symfony/polyfill-intl-icu  v1.10.0 = v1.10.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring  v1.10.0 = v1.10.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php56     v1.10.0 = v1.10.0 Symfony polyfill backporting some PHP 5.6+ features to lower PHP versions
symfony/polyfill-php70     v1.10.0 = v1.10.0 Symfony polyfill backporting some PHP 7.0+ features to lower PHP versions
symfony/polyfill-util      v1.10.0 = v1.10.0 Symfony utilities for portability of PHP codes
symfony/security-acl       v3.0.1  = v3.0.1  Symfony Security Component - ACL (Access Control List)
symfony/swiftmailer-bundle v2.6.7  ~ v3.0.3  Symfony SwiftmailerBundle
symfony/symfony            v3.4.18 = v3.4.18 The Symfony PHP framework
```

#### PHP version

```
$ php -v
PHP 7.2.8 (cli) (built: Jul 21 2018 08:05:40) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
```

## Subject

I've setup a new Symfony 3.4 project with Sonata Admin and got an error `The Sonata\CoreBundle\Form\Type\ColorType class is deprecated since version 3.10 and will be removed in 4.0. Use Symfony\Component\Form\Extension\Core\Type\ColorType instead.` on container compiling.

## Steps to reproduce

```
composer create-project symfony/framework-standard-edition foo
cd foo
composer require sonata-project/admin-bundle
# configure admin bundle according to https://github.com/sonata-project/SonataAdminBundle/blob/3.x/docs/getting_started/installation.rst
composer require phpunit/phpunit --dev
./vendor/bin/phpunit
```

## Expected results

```
  1x: The Sonata\CoreBundle\Form\Type\ColorType class is deprecated since version 3.10 and will be removed in 4.0. Use Symfony\Component\Form\Extension\Core\Type\ColorType instead.
    1x in DefaultControllerTest::testIndex from Tests\AppBundle\Controller
```

## Actual results

No deprecation messages

### PR DESCRIPTION
**Pull request by tifabien**
## Subject

I am targeting this branch, because this is BC. It moves deprecation notices of `Sonata\CoreBundle\Form\Type\*` classes into their buildForm methods.

Closes #580 

## Changelog

```markdown
### Changed
- Moved deprecation notices of `Sonata\CoreBundle\Form\Type\*` classes into their buildForm methods
```

### CODE REVIEW

**Comment by tifabien** [comment]
Note that I don't know how to deal with the 3 Base* classes. I didn't get the point of the `class_alias` and the `if (false)` stuff.

File: tests/CoreBundle/Form/Type/ColorTypeTest.php [hunk 1]
```diff
@@ -32,9 +32,13 @@ public function testDeprecation()
             $this->markTestSkipped('< Symfony 3.4');
         }
 
-        $type = new ColorType();
+        $this->factory->create(ColorType::class);
     }
 
+    /**
+     * @group legacy
+     * @expectedDeprecation The Sonata\CoreBundle\Form\Type\ColorType class is deprecated since version 3.10 and will be removed in 4.0. Use Symfony\Component\Form\Extension\Core\Type\ColorType instead.
```
Comment (tifabien) [inline tests/CoreBundle/Form/Type/ColorTypeTest.php]: Do you know if there is a way to check this expectedDeprecation message without annotation (like the `TestCase::expectExceptionMessage` method)?
Actually this annotation makes the CI --prefer-lowest build fails.

**Comment by greg0ire** [comment]
> Note that I don't know how to deal with the 3 Base* classes. I didn't get the point of the class_alias and the if (false) stuff.

You're in luck, I wrote a lengthy blog post that goes through this in great detail: https://dev.to/greg0ire/how-to-deprecate-a-type-in-php-48cf

File: tests/CoreBundle/Form/Type/ColorTypeTest.php (hunk 1 above)
Comment (greg0ire) [reply tests/CoreBundle/Form/Type/ColorTypeTest.php]: I think there isn't a way to achieve this. Note that the sf `ColorType` exists only recently in sf, which means that people using an old version of Symfony should not get a deprecation, which is probably what happens here… maybe you could use `$test->markAsSkipped()` instead?

**Comment by greg0ire** [comment]
Good job! Could you please squash commit, 1, 2, 3, and 5 together, while keeping commit 4 on its own?

**Comment by tifabien** [comment]
Done

**Comment by greg0ire** [comment]
Thanks @tifabien !

**Pull request merged**